src/Integration/Laravel/Filesystem/src/AsyncAwsFilesystemAdapter.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Illuminate\Filesystem;
+
+use AsyncAws\Flysystem\S3\S3FilesystemV1;
+use AsyncAws\S3\Input\GetObjectRequest;
+use Illuminate\Filesystem\FilesystemAdapter;
+use League\Flysystem\Cached\CachedAdapter;
+
+/**
+ * A small class that override the url() function.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class AsyncAwsFilesystemAdapter extends FilesystemAdapter
+{
+    /**
+     * Return a presigned URL to the S3 object.
+     *
+     * @param string             $path
+     * @param \DateTimeInterface $expiration
+     *
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    public function temporaryUrl($path, $expiration, array $options = [])
+    {
+        if ($expiration instanceof \DateTime) {
+            $expiration = \DateTimeImmutable::createFromMutable($expiration);
+        }
+
+        if (!$expiration instanceof \DateTimeImmutable) {
+            throw new \LogicException('Second argument to AsyncAwsFilesystemAdapter::temporaryUrl() must be a \DateTimeInterface');
+        }
+
+        $adapter = $this->getFlysystemAdapter();
+        $input = new GetObjectRequest([
+            'Bucket' => $adapter->getBucket(),
+            'Key' => $adapter->applyPathPrefix($path),
+        ]);
+
+        return $adapter->getClient()->presign($input, $expiration);
+    }
+
+    /**
+     * Get the URL for the file at the given path.
+     *
+     * @param string $path
+     *
+     * @return string
+     *
+     * @throws \RuntimeException
+     */
+    public function url($path)
+    {
+        $url = $this->temporaryUrl($path, new \DateTimeImmutable('+10minutes'));
+
+        // remove all query parameters.
+        if (false === $pos = strpos($url, '?')) {
+            throw new \RuntimeException('Expected presigned URL to include a quiery string');
+        }
+
+        return substr($url, 0, $pos);
+    }
+
+    private function getFlysystemAdapter(): S3FilesystemV1
+    {
+        if (!method_exists($this->driver, 'getAdapter')) {
+            throw new \RuntimeException(sprintf('Could not call "getAdapter" of class "%s"', \get_class($this->driver)));
+        }
+
+        $adapter = $this->driver->getAdapter();
+
+        if ($adapter instanceof CachedAdapter) {
+            $adapter = $adapter->getAdapter();
```

Comment:
Is it possible to have several layers of cache? ie. `MemoryCache(FileCache(ConcreteAdapter))`

Reply:
Yes, it is. Good catch.. I copied the code from Laravel. I didnt think much